src/Caching/Impl/ModuleDatabase.cs

```diff
@@ -33,7 +33,8 @@
 namespace Microsoft.Python.Analysis.Caching {
     internal sealed class ModuleDatabase : IModuleDatabaseService {
         private readonly Dictionary<string, IDependencyProvider> _dependencies = new Dictionary<string, IDependencyProvider>();
-        private readonly object _lock = new object();
```

Comment:
Current write lock locks for all modules which is excessive. Should be possible to write modules in parallel. So it might be  better, but more complicated, to create write lock per module (via some protected dictionary). Single lock was for the early design with a single db for everything, which is not the case.

This way `FindModuleModel` may not need to lock the db. It should to check if the module is being written using the above, but otherwise it is just a file system read operation.

Reply:
Actually, I think I can eliminate this lock entirely, as it appears that LiteDB has its own file locking system and handles multiple instances of the same file (https://github.com/mbdavid/LiteDB/wiki/Concurrency).